tests/test_protocol_optimization.py

```diff
@@ -118,3 +124,118 @@ def simulate_protocol(lambdas_k, n_samples_per_window=100, seed=None):
+    # initial_lams = linspace(0,1), along a path where nonuniform lams are probably better
+    initial_num_states = 20
+
+    def initial_path(lam):
+        offset = lam * 4
+        force_constant = 2 ** (4 * lam)
+        return (offset, force_constant)
+
+    initial_lams = np.linspace(0, 1, initial_num_states)
+    initial_params = [initial_path(lam) for lam in initial_lams]
+
+    # make a pymbar test case
+    O_k = [offset for (offset, _) in initial_params]
+    K_k = [force_constant for (_, force_constant) in initial_params]
+    test_case = HarmonicOscillatorsTestCase(O_k=O_k, K_k=K_k)
+
+    # get samples, estimate free energies
+    _, u_kn, N_k, _ = test_case.sample(N_k=[100] * initial_num_states)
+    mbar = MBAR(u_kn, N_k)
+    f_k = mbar.f_k
+
+    # make a fast d(lam_i, lam_j) fxn
+    overlap_dist = make_approx_overlap_distance_fxn(initial_lams, u_kn, f_k, N_k)
+
+    print("initial protocol: ")
+    _ = summarize_protocol(initial_lams, overlap_dist)
+
+    # optimize for a target neighbor distance
+    target_overlap = 0.1
+    target_dist = 1 - target_overlap
+    print(f"optimized with target dist = {target_dist} (target overlap = {target_overlap}):")
+    greedy_prot = greedily_optimize_protocol(overlap_dist, target_dist)
+    greedy_nbr_dist = summarize_protocol(greedy_prot, overlap_dist)
+    assert np.max(greedy_nbr_dist) <= target_dist + 1e-5  # a little wiggle room
+
+    # optimize for a target num states
+    target_num_states = len(greedy_prot)
+    print(f"optimized with target num_states = {target_num_states}:")
+    optimized_protocol = produce_target_number_of_equidistant_states(
+        overlap_dist, target_num_states, max_num_states=len(greedy_prot)
+    )
+    optimized_nbr_dist = summarize_protocol(optimized_protocol, overlap_dist)
+
+    assert len(optimized_protocol) <= target_num_states
+    assert np.max(optimized_nbr_dist) <= np.max(greedy_nbr_dist)
+
+    # also, sanity-check the overlap_dist function: d(x,x)==0, d(x,y)==d(y,x), x<y<z => d(x,y)<d(x,z)
+    rng = np.random.default_rng(2024)
+    random_lams = rng.uniform(0, 1, 20)
+    self_distances = [overlap_dist(lam, lam) for lam in random_lams]
+    np.testing.assert_allclose(self_distances, 0.0, atol=1e-10)
+    for lam_i in random_lams:
+        # compare to some random lam_j between lam_i and 1
+        lams_above = sorted(set(rng.uniform(lam_i, 1, 5)))
+        distances_to_sorted_larger_lams = np.array([overlap_dist(lam_i, lam_j) for lam_j in lams_above])
+
+        # assert x < y < z implies d(x, y) <= d(x, z)
+        eps = 1e-3
+        np.testing.assert_array_less(distances_to_sorted_larger_lams[:-1], distances_to_sorted_larger_lams[1:] + eps)
+
+        # assert symmetric
+        _distances_flipped = np.array([overlap_dist(lam_j, lam_i) for lam_j in lams_above])
+        np.testing.assert_allclose(distances_to_sorted_larger_lams, _distances_flipped)
+
+
+@pytest.mark.nightly(reason="Slow")
+def test_greedy_overlap_on_st_vacuum():
```

Comment:
@proteneer : usage example for computing an approximately minimized protocol length, given a target threshold for neighbor overlap.

(uses the pre-existing function `greedily_optimize_protocol`, in combination with newly introduced utility functions for cheaply approximating `overlap(lam_a, lam_b)` given `bisection_result`)

locally, prints something like the following:

```
Bisection iteration 0 (of 46): Current minimum BAR overlap 7.88e-45 <= 0.667 between states at λ=0 and λ=1. Sampling new state at λ=0.5…
bootstrapped errors non-normal: NormaltestResult(statistic=16.548048467699868, pvalue=0.00025505681496211193)
Bisection iteration 1 (of 46): Current minimum BAR overlap 6.34e-10 <= 0.667 between states at λ=0.5 and λ=1. Sampling new state at λ=0.75…
Bisection iteration 2 (of 46): Current minimum BAR overlap 1.79e-08 <= 0.667 between states at λ=0 and λ=0.5. Sampling new state at λ=0.25…
Bisection iteration 3 (of 46): Current minimum BAR overlap 0.00916 <= 0.667 between states at λ=0 and λ=0.25. Sampling new state at λ=0.125…
Bisection iteration 4 (of 46): Current minimum BAR overlap 0.0182 <= 0.667 between states at λ=0.25 and λ=0.5. Sampling new state at λ=0.375…
Bisection iteration 5 (of 46): Current minimum BAR overlap 0.0235 <= 0.667 between states at λ=0.75 and λ=1. Sampling new state at λ=0.875…
Bisection iteration 6 (of 46): Current minimum BAR overlap 0.0309 <= 0.667 between states at λ=0.5 and λ=0.75. Sampling new state at λ=0.625…
Bisection iteration 7 (of 46): Current minimum BAR overlap 0.153 <= 0.667 between states at λ=0.125 and λ=0.25. Sampling new state at λ=0.188…
Bisection iteration 8 (of 46): Current minimum BAR overlap 0.222 <= 0.667 between states at λ=0.25 and λ=0.375. Sampling new state at λ=0.312…
Bisection iteration 9 (of 46): Current minimum BAR overlap 0.28 <= 0.667 between states at λ=0.375 and λ=0.5. Sampling new state at λ=0.438…
Bisection iteration 10 (of 46): Current minimum BAR overlap 0.298 <= 0.667 between states at λ=0.875 and λ=1. Sampling new state at λ=0.938…
Bisection iteration 11 (of 46): Current minimum BAR overlap 0.303 <= 0.667 between states at λ=0 and λ=0.125. Sampling new state at λ=0.0625…
Bisection iteration 12 (of 46): Current minimum BAR overlap 0.333 <= 0.667 between states at λ=0.75 and λ=0.875. Sampling new state at λ=0.812…
Bisection iteration 13 (of 46): Current minimum BAR overlap 0.351 <= 0.667 between states at λ=0.5 and λ=0.625. Sampling new state at λ=0.562…
Bisection iteration 14 (of 46): Current minimum BAR overlap 0.362 <= 0.667 between states at λ=0.625 and λ=0.75. Sampling new state at λ=0.688…
Bisection iteration 15 (of 46): Current minimum BAR overlap 0.518 <= 0.667 between states at λ=0.125 and λ=0.188. Sampling new state at λ=0.156…
Bisection iteration 16 (of 46): Current minimum BAR overlap 0.565 <= 0.667 between states at λ=0.312 and λ=0.375. Sampling new state at λ=0.344…
Bisection iteration 17 (of 46): Current minimum BAR overlap 0.599 <= 0.667 between states at λ=0.188 and λ=0.25. Sampling new state at λ=0.219…
Bisection iteration 18 (of 46): Current minimum BAR overlap 0.623 <= 0.667 between states at λ=0.0625 and λ=0.125. Sampling new state at λ=0.0938…
Bisection iteration 19 (of 46): Current minimum BAR overlap 0.647 <= 0.667 between states at λ=0.375 and λ=0.438. Sampling new state at λ=0.406…
Bisection iteration 20 (of 46): Current minimum BAR overlap 0.658 <= 0.667 between states at λ=0.25 and λ=0.312. Sampling new state at λ=0.281…
All BAR overlaps exceed min_overlap=0.6666666666666666. Returning after 21 iterations.
initial protocol (from bisection): 
        # states = 23, min(d(i,i+1)) = 0.092, max(d(i,i+1)) = 0.294
optimized with target dist = 0.95 (target overlap = 0.05):
        # states = 6, min(d(i,i+1)) = 0.855, max(d(i,i+1)) = 0.950
optimized with target dist = 0.9 (target overlap = 0.1):
        # states = 7, min(d(i,i+1)) = 0.622, max(d(i,i+1)) = 0.900
optimized with target dist = 0.8 (target overlap = 0.2):
        # states = 8, min(d(i,i+1)) = 0.782, max(d(i,i+1)) = 0.800
optimized with target dist = 0.6 (target overlap = 0.4):
        # states = 11, min(d(i,i+1)) = 0.391, max(d(i,i+1)) = 0.600
optimized with target dist = 0.33399999999999996 (target overlap = 0.666):
        # states = 17, min(d(i,i+1)) = 0.110, max(d(i,i+1)) = 0.334
```